pkg/api/metrics.go

```diff
@@ -44,6 +34,38 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext) response.Response {
 	return toJsonStreamingResponse(resp)
 }
 
+func (hs *HTTPServer) handleQueryMetricsError(err error) *response.NormalResponse {
+	if errors.Is(err, models.ErrDataSourceAccessDenied) {
+		return response.Error(http.StatusForbidden, "Access denied to data source", err)
+	}
+	if errors.Is(err, models.ErrDataSourceNotFound) {
+		return response.Error(http.StatusNotFound, "Data source not found", err)
+	}
+	var badQuery *query.ErrBadQuery
+	if errors.As(err, &badQuery) {
+		return response.Error(http.StatusBadRequest, util.Capitalize(badQuery.Message), err)
+	}
+
+	if errors.Is(err, backendplugin.ErrPluginNotRegistered) {
+		return response.Error(http.StatusNotFound, "Plugin not found", err)
+	}
+
+	if errors.Is(err, backendplugin.ErrMethodNotImplemented) {
```

Comment:
This feels like a plugin error and not grafana so maybe `502 Bad gateway` instead?